### ISSUE TITLE
Emmet-like Text Expansion to Add Blocks to a Page

### ISSUE THREAD

**Issue by mrwweb** [WordPress/gutenberg#22194]
**Is your feature request related to a problem? Please describe.**
For someone working with the block editor every day, it can be really tedious to create patterns of blocks, especially those that are nested or repetitive.

It would be really amazing if we could insert blocks using an [Emmet](https://emmet.io/)-like syntax to speed up page creation. This would build on the keyboard-friendly block "slash inserter" (typing `/{search query}` and might even be able to 

For anyone who doesn't know Emmet, it is an HTML text-expansion syntax (type some stuff and then hit `TAB`) to generate HTML. For example, `section#section-1>div.box>h2{My Heading}+p*2+img[src="image.png"]` results in this HTML:

```html
<section id="section-1">
  <div class="box">
    <h2>My Heading</h2>
    <p></p>
    <p></p>
    <img src="image.png">
  </div>
</section>
```

**Describe the solution you'd like**
I haven't given syntax a ton of thought, but here's what comes to me off the top of my head:

- It would be as close to Emmet as possible. Let's not reinvent the wheel.
- It would be nice if HTML elements could work for any block that doesn't output a `wp-block-` class on the front end (off the top of my head that would be headings, horizontal rule, and paragraph)
- All other core blocks could use the block slug without the `core/` prefix for core blocks
- All other blocks would need their prefix (e.g. `ab-blocks/accordion`)
- Use Emmet's CSS class syntax for specifying block style variations, font sizes, colors, etc. and dump anything that's not recognized into the Advanced CSS Classes field.
- Possibly build directly off the block inserter, so something like `/{block expansion string}` > `ENTER` adds the blocks.

Some theoretical examples:

_Example 1_

Code: `/columns>column*3>h3{Name}+image+social-icons+p{Job Title}`

Output: 3 equal-width columns, each containing the same 4 blocks: A heading 3 that says "Name", image, social icons, and a paragraph that says "Job Title".

_Example 2_

Code: `/media-text.alignwide>quote.is-style-fancy+buttons`

Output: Media & Text block with Wide alignment. The content area contains a block with the "Fancy" style variation followed by a buttons block.

_Example 3_

Code: `/group#section-$*2>h2+media-text.has-green-background-color.has-media-on-the-right`

Output: Two Group blocks with IDs "section-1" and "section-2". Each group block contains a Heading 2 followed by a Media & Text block with media on the right and the "green" background color.

**Describe alternatives you've considered**
There are some workflows that are helpful for recreating an existing layout like Cloning a draft page, the in-development block templates, reusable blocks, and the duplicate block field, but those don't help with the initial creation of templates and also only help for situations that are repeated.

**Comment by talldan** [comment]
Yes! I really like this idea, and I've had similar thoughts. I think it would also be handy as a way to bypass the placeholder for some blocks.

I had a slightly simpler syntax in mind. Some examples I considered:
- `/table 4x4` < inserts a table with a preset size without showing the placeholder
- `/columns 30/70` < inserts a columns block with a 30/70 layout
- `/heading h4` < inserts a heading already set to heading level 4
- `/spacer 40px` < inserts a spacer with a preset height

But classnames and ids could also work if the block supports them as they're quite straightforward patterns to identify in text.

It'd be interesting to see if the autocomplete/placeholder when typing could help with this.

The feature could potentially be exposed as a type of block transform.

**Comment by mrwweb** [comment]
>  handy as a way to bypass the placeholder for some blocks.

You just clarified the unspoken reason I really want these! Thanks for pointing that out.

> /columns 30/70

Emmet uses braces for attributes, which are similar to the values in your examples, so maybe that is a good way to go? That would be easy to combine with other things too.

- `/columns[30/70].alignwide`
- More verbose but maybe clearer: `/columns.alignwide>column[30]+column[70]`
- `/heading[4]#heading-1`

(I still would love to just be able to use HTML for simple blocks!)

**Comment by draganescu** [comment]
This looks like a super nice power user feature and a great job for a plugin. I fear layout building from text commands might introduce some new kind of maintenance serving a small audience.

**Comment by Mamaduka** [comment]
Closing this as "maybe later". I think the same functionality can be explored in a plugin using transformations or autocompleters.
